portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -42,26 +44,35 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         Query query = new Query("Comment");
         PreparedQuery results = datastore.prepare(query);
         
-        List<String> texts = new ArrayList<>();
+        List<String[]> cmts = new ArrayList<>();
         for (Entity entity : results.asIterable()) {
-            String text = (String) entity.getProperty("text");
-            texts.add(text);    
+            String[] pair = {(String) entity.getProperty("email"),
+                             (String) entity.getProperty("text")};
+            cmts.add(pair);    
         }
         
         Gson gson = new Gson();
-        String json = gson.toJson(texts);
+        String json = gson.toJson(cmts);
         response.setContentType("application/json;");
         response.getWriter().println(json);
     }
 
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/shoutbox");
```

Comment:
This (`"/shoutbox"`) is dubious, because I don't see any servlet in your app src directory, having `@WebSerlet("/shoutbox")` annotation. Did you try deploying your current changes to the live server (`mvn appengine:update`) and see how it works when you access your live server from the incognito mode or from a guest browsing session? I suspect it'll say 404 (not found) in that case. Note that with your devserver alone, this case (when there's no logged-in user) can't be tested, because the devserver is accessible only when you are logged in to the cloud shell already.

A possible approach to fix this is that you just redirect to `/index.html` here as well. However, you can make a portion of your `index.html`, which will be dynamically generated by some Javascript code. It'll fetch the `/login` servlet result (just like your `getComments()` function), but it'll show a login link if the bool value from the `/login` servlet is false. If it's true, you can show some greeting message, but since your `/login` servlet just returns a boolean value, not an email address, your greeting will be limited (that is, you can't call the logged-in visitor with their email address or anything). You may fix this by making your `/login` servlet return not just a boolean value, but an email address (or possibly something more, like a name, if it's stored/available), and let the caller of the servlet (your Javascript functions) use the returned email address when composing the greeting message.